### ISSUE TITLE
`torchvision.transforms.v2.functional.resize` results differ by platform

### ISSUE THREAD

**Issue by shin04** [pytorch/vision#9122]
### 🐛 Describe the bug

The result of `torchvision.transforms.v2.functional.resize` changes depending on where the script is executed.

When we ran the container image containing the process that performs resize in different environments, the result of resize seemed to be different.
The following three environments were used to run the container, and the results on cloud run and in the local environment (macos apple silicon) were identical.
- GCE Instance (OS: Debian GNU/Linux 11, CPU: Intel Broadwell)
- macbook pro (OS: Ventura 13.2.1, CPU: Apple M2 Pro)
- Cloud Run Service 
All of them use the same container image.

GCE instance and macbook pro have identical results without running in a container, which is different for each.

A simplified script for reproduction is shown below.

```python
import hashlib

import torch
from torchvision.transforms.v2.functional import InterpolationMode, resize

torch.manual_seed(42)

random_tensor = torch.randint(0, 256, (3, 500, 500), dtype=torch.uint8)
resized_tensor = resize(random_tensor, (288, 288), interpolation=InterpolationMode.BILINEAR, antialias=True)

print(f"random_tensor hash: {hashlib.sha256(random_tensor.numpy().tobytes()).hexdigest()}")
print(f"resized_tensor hash: {hashlib.sha256(resized_tensor.numpy().tobytes()).hexdigest()}")
```

output:
```
# on gce instance
random_tensor hash: 48a3c64f637d87c72d9b9f83c84f10e343e177aa9212c820140c0ef6d2e8c73b
resized_tensor hash: b7cbf6bd095c75955cccbe0b21ca7a5525aafdf395207b85f334090e497414db

# on macbook pro
random_tensor hash: 48a3c64f637d87c72d9b9f83c84f10e343e177aa9212c820140c0ef6d2e8c73b
resized_tensor hash: 297d69b7f891efebf70002d7924d86bf79b5c621705d6aac900efb3f34752d55
```

We have confirmed that the hash value of `random_tensor` matches.

I would like to get a close result if possible, as it will considerably affect the inference process.
If anyone has experienced a similar event or has an idea as to the cause, I would appreciate your advice.

### Versions

<details>
<summary>on container</summary>

PyTorch version: 2.6.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 12 (bookworm) (x86_64)
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.36

Python version: 3.12.11 (main, Jun 10 2025, 23:56:19) [GCC 12.2.0] (64-bit runtime)
Python platform: Linux-5.10.0-35-cloud-amd64-x86_64-with-glibc2.36
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               4
On-line CPU(s) list:                  0-3
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
CPU family:                           6
Model:                                79
Thread(s) per core:                   2
Core(s) per socket:                   2
Socket(s):                            1
Stepping:                             0
BogoMIPS:                             4400.33
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            64 KiB (2 instances)
L1i cache:                            64 KiB (2 instances)
L2 cache:                             512 KiB (2 instances)
L3 cache:                             55 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-3
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; Clear CPU buffers; SMT Host state unknown

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
</details>

<details>
<summary>vm instance</summary>

PyTorch version: 2.6.0+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.18.4
Libc version: glibc-2.31

Python version: 3.13.5 (main, Jun 12 2025, 12:40:22) [Clang 20.1.4 ] (64-bit runtime)
Python platform: Linux-5.10.0-35-cloud-amd64-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 48 bits virtual
CPU(s):                               4
On-line CPU(s) list:                  0-3
Thread(s) per core:                   2
Core(s) per socket:                   2
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                79
Model name:                           Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:                             0
CPU MHz:                              2200.166
BogoMIPS:                             4400.33
Hypervisor vendor:                    KVM
Virtualization type:                  full
L1d cache:                            64 KiB
L1i cache:                            64 KiB
L2 cache:                             512 KiB
L3 cache:                             55 MiB
NUMA node0 CPU(s):                    0-3
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Mitigation; PTE Inversion
Vulnerability Mds:                    Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:               Mitigation; PTI
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Mitigation; Clear CPU buffers; SMT Host state unknown
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] Could not collect
[conda] numpy                                       2.1.3            pypi_0              pypi
</details>

<details>
<summary>macbook pro</summary>

PyTorch version: 2.6.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 4.0.2
Libc version: N/A

Python version: 3.13.1 (main, Jan 14 2025, 23:31:50) [Clang 19.1.6 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect
</details>

**Comment by NicolasHug** [comment]
Hi @shin04 , this can be caused by a number of things, instead of checking the hashes could you share  comparison of the ouptut tensors with max abs diff, mean abs diff, and other relevant metrics?

**Comment by shin04** [comment]
@NicolasHug 
Thanks for your comment.
The error in the image after resizing is shown below.

```bash
max-abs-diff: 1.0
mean-abs-diff: 0.09725838899612427
MSE: 0.09725838899612427
PSNR: 58.25153350830078
```

There is no significant divergence between the two images, but the subsequent inference results in a non-negligible difference.

**Comment by NicolasHug** [comment]
Thanks for the reply @shin04 . A diff of 1 on 1% of the pixels doesn't seem unrealistic. Can you share more about the inference results being non-negligible? Are we sure those can be attributed to the resizing?

**Comment by shin04** [comment]
@NicolasHug 
Sorry for the delay in replying.

> A diff of 1 on 1% of the pixels doesn't seem unrealistic.

I knew it...😭

> Are we sure those can be attributed to the resizing?

We save the resized tensor in each environment.
I inferred used that tensor in a different environment than the one in which I resized.
As a result, we obtained the same value as the result of the inference in the environment where the resize was excuted, so we believe that the result was affected by the resize.

**Comment by NicolasHug** [comment]
Thanks for confirming and for trying the attribution experiment @shin04 

This is a bit surprising to me. It's largely known that inference accuracy can drop significantly if there are algorithmic differences with the resize that was used during training, e.g. different interpolation mode (bicubic vs bilinear vs ...), and whether antialias was applied. These usually lead to much higher pixel difference value as well.

But I had never encountered an inference perf drop due to platform differences, especially when they're that small. Out of curiosity can you quantify the inference perf frop you're oberving?

**Comment by shin04** [comment]
@NicolasHug 
Thank you so much for your response and I apologize for the delay in replying each time.

This is the result of comparing inference results for 1000 same images:

```
MAE: 0.00252523
RMSE: 0.00505667
Maximum absolute diff: 0.05414611
```

I have never experienced anything like this and a bit confused. Should we allow for this level of error...?

By the way, I have confirmed that there is no difference in the settings for completion algorithms, anti-aliasing, etc.

**Comment by NicolasHug** [comment]
@shin04 can you clarify what the MAE and RMSE represent here? Are these the output of the resize? I'm interesting in knowing the effect that this may have on the **model prediction* performance, e.g. measuring the accuracy if you're doing classification, or any other model prediction metric that you're using?
